stats/opentelemetry/client_metrics.go

```diff
@@ -0,0 +1,262 @@
+			return "other"
+		}
+	}
+	for _, opt := range opts {
+		if _, ok := opt.(grpc.StaticMethodCallOption); ok {
+			return method
+		}
+	}
+	return "other"
+}
+
+func (csh *clientStatsHandler) streamInterceptor(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	ci := &callInfo{
+		target: csh.determineTarget(cc),
+		method: csh.determineMethod(method, opts...),
+	}
+	ctx = setCallInfo(ctx, ci)
+	startTime := time.Now()
+
+	callback := func(err error) {
+		csh.perCallMetrics(ctx, err, startTime, ci)
+	}
+	opts = append([]grpc.CallOption{grpc.OnFinish(callback)}, opts...)
+	s, err := streamer(ctx, desc, cc, method, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (csh *clientStatsHandler) perCallMetrics(ctx context.Context, err error, startTime time.Time, ci *callInfo) {
+	s := status.Convert(err)
+	callLatency := float64(time.Since(startTime)) / float64(time.Second)
+	if csh.registeredMetrics.clientCallDuration != nil {
+		csh.registeredMetrics.clientCallDuration.Record(ctx, callLatency, metric.WithAttributes(attribute.String("grpc.method", removeLeadingSlash(ci.method)), attribute.String("grpc.target", ci.target), attribute.String("grpc.status", canonicalString(s.Code())))) // needs method target and status should I persist this?
+	}
+}
+
+// TagConn exists to satisfy stats.Handler.
+func (csh *clientStatsHandler) TagConn(ctx context.Context, _ *stats.ConnTagInfo) context.Context {
+	return ctx
+}
+
+// HandleConn exists to satisfy stats.Handler.
+func (csh *clientStatsHandler) HandleConn(context.Context, stats.ConnStats) {}
+
+// TagRPC implements per RPC attempt context management.
+func (csh *clientStatsHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	mi := &metricsInfo{ // populates information about RPC start.
+		startTime: time.Now(),
+	}
+	ri := &rpcInfo{
+		mi: mi,
+	}
+	return setRPCInfo(ctx, ri)
+}
+
+func (csh *clientStatsHandler) HandleRPC(ctx context.Context, rs stats.RPCStats) {
+	ri := getRPCInfo(ctx)
+	if ri == nil {
+		// Shouldn't happen because TagRPC populates this information.
+		return
+	}
+	csh.processRPCEvent(ctx, rs, ri.mi)
+}
+
+func (csh *clientStatsHandler) processRPCEvent(ctx context.Context, s stats.RPCStats, mi *metricsInfo) {
+	switch st := s.(type) {
+	case *stats.InHeader, *stats.OutHeader, *stats.InTrailer, *stats.OutTrailer:
+	case *stats.Begin:
+		ci := getCallInfo(ctx)
+		if ci == nil {
+			// Shouldn't happen, set by interceptor, defensive programming. Log it won't record?
+			return
+		}
+
+		if csh.registeredMetrics.clientAttemptStarted != nil {
+			csh.registeredMetrics.clientAttemptStarted.Add(ctx, 1, metric.WithAttributes(attribute.String("grpc.method", removeLeadingSlash(ci.method)), attribute.String("grpc.target", ci.target))) // Add records a change to the counter...attributeset for efficiency
```

Comment:
attributeset?

Reply:
Sorry, note to myself, forgot to delete. But yeah, attribute set is slight more efficient haha :).